File: 2025/day02/ruby/solution.rb
#!/usr/bin/env ruby

# Read input file
input_text = File.read(File.join(__dir__, '..', 'input.txt')).strip

def is_invalid_id_part1(num)
  """Check if a number is invalid (a pattern repeated exactly twice)."""
  s = num.to_s
  length = s.length

  # Must have even length to be repeated twice
  return false if length % 2 != 0

  # Check if it starts with 0 (leading zeros not allowed)
  return false if s[0] == '0'

  # Split in half and check if both halves are identical
  mid = length / 2
  first_half = s[0...mid]
  second_half = s[mid..-1]

  first_half == second_half
end

def is_invalid_id_part2(num)
  """Check if a number is invalid (a pattern repeated at least twice)."""
  s = num.to_s
  length = s.length

  # Check if it starts with 0 (leading zeros not allowed)
  return false if s[0] == '0'

  # Try all possible pattern lengths from 1 to length//2
  # The pattern must be repeated at least twice, so max pattern length is length//2
  (1..length / 2).each do |pattern_length|
    # Check if the string length is divisible by pattern_length
    if length % pattern_length == 0
      pattern = s[0...pattern_length]
      # Check if repeating the pattern gives us the original string
      return true if pattern * (length / pattern_length) == s
    end
  end

  false
end

def part1(input_text)
  # Parse ranges from input
  ranges = []
  input_text.split(',').each do |part|
    part = part.strip
    if part.include?('-')
      parts = part.split('-')
      if parts.length == 2
        start_val = parts[0].to_i
        end_val = parts[1].to_i
        ranges << [start_val, end_val]
      end
    end
  end

  total = 0
  ranges.each do |start_val, end_val|
    (start_val..end_val).each do |num|
      total += num if is_invalid_id_part1(num)
    end
  end

  total
end

def part2(input_text)
  # Parse ranges from input
  ranges = []
  input_text.split(',').each do |part|
    part = part.strip
    if part.include?('-')
      parts = part.split('-')
      if parts.length == 2
        start_val = parts[0].to_i
        end_val = parts[1].to_i
        ranges << [start_val, end_val]
      end
    end
  end

  total = 0
  ranges.each do |start_val, end_val|
    (start_val..end_val).each do |num|
      total += num if is_invalid_id_part2(num)
    end
  end

  total
end

puts "Part 1: #{part1(input_text)}"
puts "Part 2: #{part2(input_text)}"

File: 2025/day02/python/solution.py
from pathlib import Path

input_text = (Path(__file__).parent.parent / "input.txt").read_text().strip()

# Parse input
lines = input_text.split("\n")


def part1():
    def is_invalid_id(num: int) -> bool:
        """Check if a number is invalid (a pattern repeated twice)."""
        s = str(num)
        length = len(s)

        # Must have even length to be repeated twice
        if length % 2 != 0:
            return False

        # Check if it starts with 0 (leading zeros not allowed)
        if s[0] == '0':
            return False

        # Split in half and check if both halves are identical
        mid = length // 2
        first_half = s[:mid]
        second_half = s[mid:]

        return first_half == second_half

    # Parse ranges from input
    ranges = []
    for part in input_text.split(','):
        part = part.strip()
        if '-' in part:
            # Split on the last dash to handle negative numbers (though not in this input)
            # Actually, we need to find where the dash is that separates start from end
            # Since ranges are like "11-22", we split on dash
            parts = part.split('-')
            # For ranges like "1188511880-1188511890", there's only one dash
            if len(parts) == 2:
                start, end = int(parts[0]), int(parts[1])
                ranges.append((start, end))

    total = 0
    for start, end in ranges:
        for num in range(start, end + 1):
            if is_invalid_id(num):
                total += num

    return total


def part2():
    def is_invalid_id(num: int) -> bool:
        """Check if a number is invalid (a pattern repeated at least twice)."""
        s = str(num)
        length = len(s)

        # Check if it starts with 0 (leading zeros not allowed)
        if s[0] == '0':
            return False

        # Try all possible pattern lengths from 1 to length//2
        # The pattern must be repeated at least twice, so max pattern length is length//2
        for pattern_length in range(1, length // 2 + 1):
            # Check if the string length is divisible by pattern_length
            if length % pattern_length == 0:
                pattern = s[:pattern_length]
                # Check if repeating the pattern gives us the original string
                if pattern * (length // pattern_length) == s:
                    return True

        return False

    # Parse ranges from input
    ranges = []
    for part in input_text.split(','):
        part = part.strip()
        if '-' in part:
            parts = part.split('-')
            if len(parts) == 2:
                start, end = int(parts[0]), int(parts[1])
                ranges.append((start, end))

    total = 0
    for start, end in ranges:
        for num in range(start, end + 1):
            if is_invalid_id(num):
                total += num

    return total


if __name__ == "__main__":
    print(f"Part 1: {part1()}")
    print(f"Part 2: {part2()}")
